### ISSUE TITLE
Underpathing object fills.

### ISSUE THREAD

**Issue by carolundy** [inkstitch/inkstitch#257]
Excuse me if you are already working on this, but I would really like to see this option in the parameters, rather than the jump stitch option. Pathing around the edges to fill irregular shapes is ok if the edge is to be covered with a border, or if the fill is not dense, in which case underpathing would show. Where there is no border though, I would prefer to have the pathing underneath the stitching. I have attached a short video, which is a screen capture of the bernina software stitch simulator, demonstrating what I mean. 
I gather some complicated algorithm is necessary to do this.
I used a single angle of 45degrees and no underlay, to simplify. If there was an underlay, there would need to be an additional path created from the end point of the underlay to the start point of the object.
I don't want to insult you clever mathematicians and coders, but it appears the break up of the object is defined by the stitch angle. The stitch order of the various segments seems to start with the object furthest from the end point. Paths are created wherever the stitch path, if continued, would cut off access to a segment.
[under-pathing-bernina-example.zip](https://github.com/inkstitch/inkstitch/files/2247866/under-pathing-bernina-example.zip)
Cheers Carol

**Comment by tatarize** [comment]
The core algorithm used in inkstitch is a Eulerian fill (or something I'm going to call a Eulerian fill as a copy/improve the feature for my own uses) which uses the edges as scaffolding. It's based on some work for reasonable approximations of travelling salesman problems with regard to pathing. The goal of the algorithm is, mostly to absolutely minimize the amount of extra stitches used to complete the fill. But, one of the major advantages is that it necessarily works for all fills, with or without holes, in any shape and without consideration of the shape in question.

What Bernina and many other software packages do is a  much weaker algorithm akin to a flood fill. It does have some advantages, like because you calculated all the stuff you needed you can use that stuff.  But, a great many more disadvantages dealing with the edge conditions and other fill conditions. This does however mean that the core algorithm behind both of these fills is fundamentally different.

Inkstitch goes around the edges because fundamentally that's what the algorithm tends to require. At least as it is currently implemented. There's a few ways to adapt this though with some advantages and mostly disadvantages.

* Force the Eulerian graph to have underpath segments that would get necessarily consumed when they are covered. They would be loops between rungs and if those rungs are used it would automatically drop those loops. Allowing the underpathing to be used some particular cases, going the correct direction.
* Draw the graph in a fundamentally different overtly using underpathing rather than outerpathing. This would basically require single directional underpathing and loops between montone objects. This would require that unless the exit point is at the tip of the object you would be required to take that path into the object and then fill out of the path. 
* Optimize the pathing around the edge to consider whether a straighter path would occupy space that would later be covered by stitches and if so, draw a straighter line through this area.

The last one seems like the most implementable. Many of these versions suffer from some fundamental problems like monodirectional scaffolding (a big no-no for Eulerian graphs) or basically requires creating the original algorithm and reintroducing the problems it produced, and consequently limiting the power of the algorithm for future stitches etc. 

The way the Bernina algorithm works is that it will fill the object in one direction until it detects that the area above the current area suddenly exists, like there's a split node and a new area. It will then go and find the one of the extremes, this means either travel in the shape all the way to the bottom or to the new top and to do this with underpathing. Then continue the algorithm with the same allowances. This is the way a flood fill works except when it finds a point where it hasn't filled a section above it, it travels to the topmost splitnode with underpathing. Though the Bernina video clearly shows it travelling to the a merge node, at the bottom.

The way the Inkstitch algorithm works is that it takes any shape and then a fill that will occupy shape and uses the shape as a cutter for the fill. It then travels around the edge adding straight edge nodes where two adjacent intersections between the shape and the fill occur, which adds one piece of edge for each of these than adds another piece of edge for every other section around the edge. It then uses a variety of heuristics to find a Eulerian path through this graph that maximizes local boustrophedonic (as the ox plows) travel. 

The massive advantage to inkstitch's algorithm is that if you replace those heuristics with something less restrictive and you can fill any object with any fill. You can naively do wavy lines or loops, or any sorts of modifications of fills. Eventually those will be free. They will just be natively implementable and solvable in the algorithm that inkstitch currently uses. Underpathing however, is really only effectively calculated using a flood fill-like algorithm. Where you detect areas above the fill line because you look for those then underpath to the topmost area within that. There is clearly a need to hide stitches inside the area being covered but I don't yet see an easy way to do that given the algorithm being used.

I would however, expect that inset the edge line when you are consuming virgin edge is a perfectly workable feature. It would mean that if you are traveling through an edge line where the nodes have never yet been visited that you could migrate the line you are travelling towards the inside of the shape. Especially along concentric line paths.  This would achieve some of the desired feature without requiring any fundamental change to the algorithm. And further wouldn't start screwing up for other types of fills. 

Add a checkbox to: travel along a concentric line within the shape if the edge you are traveling along has not been visited.

**Comment by lexelby** [comment]
Thanks for opening this issue.  I'd really love to have underpathing, too!

The simple truth is that I've just never figured out how to do it.  I could never get the "flood fill" algorithm that @tatarize described to work reliably for all kinds of shapes.  It's also not clear to me how one might choose to lay out those meandering underpaths.

As @tatarize suggests, the current algorithm, for all its flexibility, may not be well-suited to underpathing.  It's not immediately clear to me how we might adapt it or rewrite it to get underpathing.  I'll have to think on it.

I know it's definitely not what you're looking for, but you might get at least some benefit from the underlay inset option.  Then at least the "outerpathing" for the underlay won't show.

**Comment by lexelby** [comment]
@tatarize How about we think along the lines of your 2-opt idea?

Consider a situation where auto-fill does back-and-forth rows down to the bottom of a shape, and then running stitch along the outside to get back to where it started.  We could instead do underpathing to get to the bottom and then back-and-forth to get back.  In other words, just like 2-opt, we reverse the paths to make the running stitch come first, then inset it so that it gets covered up.

This operation should always be "safe".  Any time we find a sub-loop that ends with running stitch along the outside, we can reverse it to get the running stitch to come first, then inset it to hide it.

I wonder if applying this repeatedly will always be successful, or if there are certain kinds of "outerpathing" that can't be treated in this way?

**Comment by tatarize** [comment]
Given some amount of insetting it's always the case that that a path must exist. Since the minimum inset is 0, and this is already identical to what's currently done, then insetting must always be possible even if you must vary it to avoid tangental strikes.


![underpath](https://user-images.githubusercontent.com/3302478/43562365-f714a0ee-95d0-11e8-9920-d13d9abb0ccd.png)

Note that replacing the blue path with the red path could hit a rung even though no node along the blue path had been used yet.

---

The way you are suggesting doing this seems to hijack the 2-opt to serve that purpose exclusively. It should rather be a weigh on the heuristic if done at all. 

I entirely think the post processing step would be nice, namely if it is situation 1 do that inset. But if it is situation 2 perform no changes. Then simply seek to maximize state 1 during the 2-opt. Rather than turn all instances of state 2 into state 1.  Obviously 100% of the time, is certainly maximal. But, you might be trading too much for the end result, though I certainly can't prove it. However I think I can prove that it will be true that there is a solution. 

You'd have a bunch of different paths that would develop underpathing and these areas would be monotone sections within the fill. In fact, one could effectively show if one broke a given shape into monotone parts you could certainly create each montone section as a monodirectional loop and since any monotone section would only ever require one underpathing event there must be a solution to this. Since you can show that all monotone sections must also be loops you can show that there must be a path that connections all the nodes while also being permitted to preserve directionality within these loops. If you can partition a shape into monotone sections and you can make this montone section a loop (by adding a connection from the end to the start) it seems you must be able to solve the circuit regardless the required direction within the loops. Since you cannot force overall contradictory directions, within the same loop without being part of a subloop. So yes, eventually it would certainly flip them all and replace every edge walk with an inset edge walk in the correct direction.

It might be entirely possible for you to do this, that you use your current heuristics solve for a path, then in those cases where you travel along the edge, just go ahead and find those, replace them with an inset version and flip that part of the loop.  I would only strongly recommend not making this iron-clad, must always be done. Rather prefer state 1 to state 2, but if you have other values these might be able to make a non-inset path preferred because a different direction allows so other priority to be met.

I'd strongly prefer folding this into my current 2-opt suggestion, rather than making it the raison d'etre of the operation. After all your well optimized first loop either through my suggested use of 2-opt or through your current heuristics might be completely screwed over if you throw in a bunch of flips as a sort of post processing step. If done entirely as one step where inset-pathing is a post-processing operation that could always exist, but you optimize selecting for paths that utilize that step more it seems like it would be preferable. It seems like you might be right though that you might be able to ensure that that optimization can exist in every case. But, I'd prefer just giving the optimization bonus points rather than a requirement.  If for no other reason than it might balance better with future unknown fills etc. My only objection to it is akin to my objection to your current heuristics, you're better off giving it a value and solving for path with the best value.  In theory using 2-opt for everything should be able to use it for everything, even this. Though here if there is a case where it would results in a suboptimal solution the algorithm should just walk the edge like it currently does in order to find the better option. 

Keep in mind I'm unable to think of a case where that might be the case, but the point of 2-opt is that you shouldn't have to. You just assign weights to the paths, and find the best one (or close-thereto). But, yeah, if you wanted to always underpath like that it might necessarily work, but at what cost!? I'd suggest actually calculating the cost and see if that's worth paying, rather than always inset and always flip.

In any case, you could certainly just do underpathing in those cases when the edge is unvisited and give the user a checkbox for that feature. Performing the operation when state 1 occurs. Even if you don't always transform state-2 into state-1 or give the optimization routine some value for doing so.

**Comment by carolundy** [comment]
Wow, I think I understand enough of what you have both said to comment. Insetting the pathing would be fine, if you can get it to work reliably. I also like the idea of having the option to choose to underpath, or travel on edges. Travel on edges will be necessary when all those pretty, open, lacy fills are added to this plug in. - LOL no pressure guys.
In the current release, insetting the underlay is something I will definitely do. On my test sew the pathing along the edges was barely visible, and I am sure, by adding some pull compensation, It would not be noticed. I did use a fairly stable fabric and stabilised it appropriately. In any event, I am feeling confident to make some videos about the software. I'm just abit busy for the next week, but will get on to that as soon as I can.
Thanks again for all your efforts.
Cheers Carol

**Comment by tatarize** [comment]
In theory if it were just a post-operation checkbox that insets the pathing into the fill in those cases where doing so would hide it, you could manually turn it off and on depending on the fill you would be using. It also has the strong advantage that you could do it regardless of the rest of the algorithm. Then optimizing for doing that more often could also be done and it wouldn't throw a wrench in anything. It also is progressive in the feature build. You could add the algorithm to perform the operation for state-1 even without optimizing the path reversing the other operations and ensuring that all the paths must be underpaths. And then do the other bits later.

But, due to the nature of the fills inkstitch uses the standard underpathing isn't available. It isn't simply a matter of stitching to the topmost point that you're already looking for because you need to find it to do more filling.  However, because inkstitch's algorithm is adapted from a solution for an approximation of the traveling salesman problem is is the case that the amount of extra thread around the edge cannot exceed 1 full tracing. It should already be really lightly applied and fairly unnoticed. And it would also be possible to cheat and force a trim and jump if this were some kind of deal breaker.  Also adding in various other fill types as will undoubtedly happen, the underpathing will become more and more problematic because the utility of the scheme depends generally on highly back and forth nature of the fill. If the fill were crosshatched or squiggly or swirly or whatnot the general form of the underpath collapses. You could still do my post-process inset variation, but most all the traditional types and the various other methods may well collapse. Without a firm demarcation between the filled and unfilled areas where you could have the entire thing half-filled underpathing becomes a sort of odd artifact of one particular way of doing things. So do Lex's heuristics. Hence some of the discussions about them potentially getting a bit hinky with other fills. And the only real way to show the superiority or the efficacy of such methods is to actually build them.

On the bright side it did spur discussion and ideas sufficient to implement something along the right lines, which is more than we had the day before.

**Comment by lexelby** [comment]
@carolundy I'm glad you're gaining confidence in Ink/Stitch!  I look forward to seeing the videos you produce.  No pressure on that of course -- this is all volunteer work.  @X3msnake can add you to the official Ink/Stitch youtube channel if you're interested.

**Comment by lexelby** [comment]
@tatarize Looks to me like your concerns come down to:

a. What if we inset so far that we run over an opposing edge?
b. Instead of post-processing to make outerpathing hideable, can we optimize for hideable underpathing in the initial Euler path generation?

I can solve **a**. Instead of insetting by a set amount, aim for the center of the rows that we're passing through.  That way, we always avoid going too far, and we stay far away from the edge in most cases.

For **b**, I see what you're saying, but I don't think this is the direction I'd want to go.  My heuristics really do work pretty well at the moment.  They generate a nice, logical-seeming path that optimizes for long runs of back-and-forth fill.  I'm not even sure how I'd go about weighting for making the outerpathing come first, but I think if we did that, we'd probably find that to be at odds with my existing weighting that produces nice fill paths.

In any case, given what you've said , I have increased confidence that the post-processing step would indeed work for all cases.  I realized that it's not even necessary to look for sub-loops and reverse them.  If there's outerpathing running along existing back-and-forth fill, then we can snip the outerpathing and back-and-forth sections and swap them, reversing one or the other as necessary.  The overall path integrity is maintained and this avoids incidentally re-reversing sections we've already reversed previously.

In short, it looks like I've unexpectedly found a solution to the underpathing problem, and I'm super-stoked. :D 

That said, what I've described here, while definitely doable, is still non-trivial to implement, so it may be awhile before we get this feature, @carolundy.  Keep your eye here for updates.

**Comment by carolundy** [comment]
That is great progress, I will patiently await the implementation.
I'll  touch base with @X3msnake when I have some videos done.
Cheers Carol

**Comment by tatarize** [comment]
You solve *a* for parallel rows. If the fill is random directions etc or some other modified fill your solution might well exist but I can't vouchsafe how it would work with every supposed future fill. It might be solved with just basic collision detection. Morph the line into progressively straighter versions until it is either a straight line or collided with something inappropriate. I'm not sure your solution solves for the most general case.

For *b*, I'd again object along the same lines. Your heuristics seem fragile and might blow up in the general case. You can easily give value to the things you value and then solve for the shortest path with 2-opt or rather the most valuable path. It's pretty fast and effective for TSP problems. 

The fill and the outerpathing might be going different places, you could certainly find a loop and swap two paths to the same end, but any path between the end of that fill section and the end of the outerpathing would need to be reversed (if they go to the same end node then that path is zero nodes long). While the odds of a loop in there are pretty high, I'm not sure you could absolutely demarcate between fill and edge in a way that provides ample guarantees that the fill is entirely contiguous. And a subtour within a non-contiguous fill could well contain another edge run. I'm not sure you could entirely avoid sub swaps. Though it's clearly always still going to be solvable.

However, it wouldn't be solvable if you try to pull this trick twice. You are totally tying your hand with regard to doing this trick for some other issues. So if you wanted a particular line going a particular direction, you could no longer guarantee that as you might end up flipping things.

Two opt (the single path swapping operation not the optimization routine as a whole) does however only take a couple lines of code, so it would entirely be doable to add the check box for the inset pathing where available and another for make sure it's always available. 

---

**Comment by tatarize** [comment]
> "I'd go about weighting for making the outerpathing come first, but I think if we did that, we'd probably find that to be at odds with my existing weighting that produces nice fill paths."

Basically for 2-opt optimization what you're doing is performing loop swaps. Each node in the entire path is nearly always going to be of order 4. Which means you can always go from the current circuit to one that swaps the loops. (a-b, c-d) can become (a-c, b-d) or (a-d, b-c) whichever one doesn't create disjointed solution. To do the optimization you run these various path switches until there is no path switch which provides a better value. This is pareto-optimal, and is going to be pretty close to optimal.

You would weigh the paths such that you don't take any cost hit or get a bonus for any path that contains loops made entirely out of scaffolding (these can always be removed so they are good to find). You give it some extra points if it takes the rungs in order. So if it went with rung 5 last time, it gets better score for using rung 6 then rung 7. This will value the highly contiguous boustrophedonic paths. You will also be able to tell whether a particular path is along a filled or unfilled edge. Because as you iterate through the path, you can tell that the edge nodes you are traveling through have not been traveled through. So any path of pure edge, gets some extra value if it travels through previously unvisited nodes. You can then give some extra value to using less scaffolding (pure distance metrics is what 2-opt usually does), or putting the scaffolding adjacent to the start or end (in which case you can start at the end of that scaffolding).

The only time it wouldn't come up with the perfect route is in the case where some of these goals come in conflict. For example preferring to loop a scaffolded edge and thus remove that scaffolding would beat travelling that edge before a fill, so it would opt to remove that edge if possible more than prefer to utilize that edge before the fill, and only if there's a pressing need for other reasons actually do an outerpathing on that. Such things might occur if it's filling a bunch of different shapes and satin fills and the shortest path requires a bit of a jump, in which case it might even be preferable to allow a trim and move, if it made everything else really really efficient. But you could just give this option a value and it would use it only if it needed it.

Such a weighing factor would be at odds with the methods you currently have, but not at odds with the methods I've previously proposed. You really could solve the entire embroidery with a single circuit that you optimize according to weights. You could make making sure you edge-path before fill one of the things you weight for. But, also keep the weights for other things like contiguous back and forth fills. You then wouldn't be tied to a lot of stuff that expects that kind of fill. And it would give you a really nice sort of one size fits all solution. And keep in mind, we're talking about TSP graphs, and the solution I'm suggesting is the foremost optimization routine to solve those. They really do sort of magically find optimal routes without regard to some objective standard of optimal. You weigh the stuff you want and it will provide that stuff. 

There's a real possibility that it will even outperform your heuristics, but it would also not require them, so you could solve any graph and wouldn't have to fear a curved line or loops or patterns or anything. You'll still get a pretty close to optimal route.

https://en.wikipedia.org/wiki/Inventor%27s_paradox

**Comment by tatarize** [comment]
Keep in mind isn't about coaxing things into maybe coming out the way you want them to, you actually just find any solution to the Eulerian circuit then optimize that. So the heuristics only occur during optimization not during building the circuit with tricks and tips to try and coax it into things you'd want. Rather you just find any solution, then optimize that circuit based on the things you want from it. You could translate some of your metrics like preferring small loops to big loops etc, but these are all part of the value routine that gives you the overall value of that particular path (in traditional TSP it's often just shortest distance, but any metric works).

**Comment by lexelby** [comment]
I think I see.  Your last comment made it click: find any eulerian path and then refine it until it suits my needs.  Neat!  I'll think it over.

If we're optimizing by doing 2-op pt loop swaps, then how do we find the best loops to swap in a reasonable amount of time?

 We don't have to limit ourselves to ac2-opt-style swap.  We can also fine any place where two parts of the path both start at point A and end at B and exchange them.

**Comment by tatarize** [comment]
We don't find the "best" loops to swap. We try swapping them all. If that swap improves things we swap it. If it doesn't we don't. We try things until we can't anymore. Yes, order might matter, but it should just flip back and forth until things fix up pretty optimally. You can do something like Lin-Kernighan to find the best loops to swap, but it's a bit overkill and the math is a bit, oof. 2-opt can be done in a few lines of code and a value routine.

2-opt is pretty fast. Usually it's the case that performing the 2-opt ends up at a point where the where no 2-opt operation will improve it the path's value in a rather short period of time. If you can calculate a path's value pretty quick and calculate a swap basically instantly you're left with hammering that until it stops improving, which does happen pretty soon (less than a second is totally doable). Which means that it has arrived at a pareto-optimal path. This very much often seems to be optimal but to prove that you'd have to brute-force an NP problem. 

You could even mix and match a bit. It takes an established path and refines it. You could by definition use that on the path you already found with your heuristics. If it could not improve on that, it would test each node for swapping and then terminate. The way it works is that it can only improve a path's value.  But, it would likely solve for better than your answer, faster than your method took to generate that path.

The optimization you're talking about is like basically 3-opt. Basically at point A and point B we will have two incoming paths and two outgoing paths, and 2 segments between them. So:

...---1---A---2---B---3---...
...---4---A---5---B---6---...

You can delete the connections within node A and B. Then try all the methods to hook them back up. These come back to identity (same way they were before), 2-opt at A, 2-opt at B, and 2-opt at A and B. 

...---1---A---2---B---3---...
...---4---A---5---B---6---... (Identical)

...---1---A---5---B---6---... 
...---4---A---2---B---3---... (2-opt at A)

...---1---A---2---B---6---...
...---4---A---5---B---3---... (2-opt at B)

...---1---A---5---B---3---... 
...---4---A---2---B---6---... (Segment swap / 2-opt at A and B)

This actually does permit some operations that might not otherwise be tried. And you can do this operation on more nodes, and 3-opt does certainly do better but it also takes more time. And in theory even 2-opt should find a really good route. 

You could do both passes. If you run out of 2-opt operations, you fall back to the 3-opt only doing the 2-opt of A and B (segment swap) as you will already know that either just A or just B will not produce anything of value (because you tested them in the 2-opt passes). You could even do a single or couple passes.  Though, since it's all NP it's best to settle a path you can't improve. You could in theory search forever for the perfect path with progressively bigger swaps, but you'll likely find the best path or something nearly as good pretty quickly. (you just won't be able to prove this fact)

And there's clearly room for improvement. for example:

![image001086](https://user-images.githubusercontent.com/3302478/43676468-ee372b54-97a6-11e8-9d22-663bb1f4336e.png)

![image002194](https://user-images.githubusercontent.com/3302478/43676462-c59b35c8-97a6-11e8-995a-26f405bf646e.png)

Hits the other side then does...

![image002304](https://user-images.githubusercontent.com/3302478/43676478-154ec332-97a7-11e8-9b3a-f1d49c724a4e.png)

That was certainly local. But, it was totally wrong. It should have edged over to the correct spot then filled all the way across rather than cutting across there. It's clear to see why it does that. But, you can also see that if it was allowed to trade sections around, it could fix that. It would swap the edge-walk-back with the fill and end up with a path that goes all the way across as that would have more value, and it would take a simple swap. Even a single pass could fix a lot of these issues and outperform your current heuristics. In fact, if you consider any parts you don't like with the stuff it does, you can fix most of them with a loop flip.

If you build something based on tricks and just try to coax it to doing the right thing you can get some good results but it will be really fragile, and you could literally just solve for the whatever you're calling a good result, and you'll have a faster and better solution, moreover it would be far more robust and would work for any type of fill, which isn't something sneeze at considering the inherent power behind Eulerian fills. You could give it a preferred order and it'll find a path that comes really close to that order without violating any requirements in the graph, and do so really quickly.

**Comment by lexelby** [comment]
Thanks, this gives me a lot to think about.

> If you can calculate a path's value pretty quick and calculate a swap basically instantly

That's the crux there.  I think we'll find that it is _not_ fast to calculate the value of a path -- at least not repeatedly.  We have two basic "bonuses" to consider: back-and-forth fill and outline travel before fill.  I'm thinking these will require walking the path while tracking visited edge vertices (in a hash table).  Probably O(n log n).

Consider that n is pretty big: a medium-sized fill has thousands of stitches.  Consider also that the path may have a huge number of loops, even after we discard scaffold-only loops.  That means we may end up at O(n^2 log n) or an even higher exponent.  I have some doubts on the speed. 

We can speed this up if we can avoid calculating the score of the full path for each swap.  Instead we'd calculate a delta.  That kind of calculation will get really complex to code, though.

**Comment by tatarize** [comment]
A raw calculation would take O(n) time. Say you decide the best thing for us it to preset an optimal order we would like things to be in. We want these rungs in this order, and these other rungs in this order, with the scaffolding as don't care for the ordering. We could certainly check of we're in optimal order as we walk the path. Giving (for example) 2 points if we're in directly sequential order and 1 point if we're at least in monotonic order. This would also mean we'd know the last_value_preferrred_order_number visited, within each node for the walk and thus could calculate scaffold loops as simply revisits to node N with the same last-rung value.

The edges themselves usually are required to store information this is easier and more direct than using a hash table. Since the hash for the edge is just looking up that exact information for where we currently are. You would be better off adding that data to the edge directly, or in an array lookup directly associated rather than hash associated.

In a great many cases you can even do the calculation in constant time, depending on the calculation. If you store the information for the current value of the route which you already know. You can calculate the change directly. When 2-opt is used to optimize distance this is pretty trivially as the only things changed are the distances of the current swap and you can calculate those and know the distances for everything else is identical. Something similar might be possible here.

To do such a thing the values would have to be symmetric you would need to value, reverse optimal order equal with direct optimal order. Then the only thing that would matter is change and whether or not that swap more or less sorts the edges. If we don't do this we'd end up with O(n) as we'd have to walk the loop to sum the values.

You have connected edges in a node, can well have the local value of this connection, but also the local value of this connection in reverse. Say we have 4 edges into node A, that really means we can know the local values for 1--A--2 and 3--A--4 as well as 2--A--1 and 4--A--3. When we flip a loop with 2-opt we are just reversing the order for a subset. In fact in the path it works like finding all the elements between element A1 and A2 and reversing their order. To reverse something like that we're usually already committed to an O(n) walk. If we do a 2-opt we'd only really be changing the local values, because rather than 1--A--2 and 3--A--4 we'd have 1--A--4 and 3--A--2, and the rest of the connections would be the same but in reversed order. For distance calculations, their overall value would be the exact same and you solve the value change in constant time by just calculating the difference in the distance. Here we'd need to calculate the local change and the value change for the reversed subset which is O(n) though quite likely we'd already take the O(n) hit to reverse the thing anyway.

We don't walk every single available path ever. We literally only ever walk the current path we have. We don't walk other paths. We have a list of n nodes. These are connected up exactly 1 way.  When we walk the path we have to walk that single path, not the tree that is available to us. The number of nodes doesn't change the time complexity. We don't pay an extra hit for traveling a loop, we go along the path, we'll hit the same node again on that walk, it'll have the data from last time we were there. If we hit the same data again and has the same number of rung count from our previous visit, that's a loop. We found it within the same walk. We'd be talking doing an entire pass in O(n^2) For each node, we could perform a 2-opt, to calculate whether this 2 opt is worth it, we walk the current walk there forwards and we walk the current walk there backwards simulating reversing that sub-path. We then know whether performing the 2-opt is better. That was O(2n) we do this for each of the N nodes and we have O(n^2). 

Also, we do not discard the scaffold-only loops until the end.  That would be a quick post processing walk in O(n) time but they can't be pre-clipped. We could in theory do a bit of an optimization to say that if there is a scaffold only loop and a node would know that, rather than going down that loop, it could just pretend it did, one assumes there'd be nothing asymmetric about them. Tell the walking routine the value for that loop and send it out the correct direction as if it walked the loop. But, we can't modify the graph. The graph is absolutely the graph. We only modify our path through the graph and that path must be a complete circuit.  Once we optimize everything we can then do the post-processing work. But everything happens in that optimization. It needs the options.

Oh, duh, you could do the same trick for *every loop*. If a node knows that it has a loop and it would because you walked it. It would know, for example, edge 2 and 4 eventually connect. So if you come from edge 1 or edge 3 you will by definition, walk the same path you walked before (or the reverse of the path you walked before) and that's literally 2 numbers which you could store. You could not only skip scaffolded loops with that trick but all loops with that trick. You'd just have to update the values for the loop walks if the loop contains a loop that was performed a 2 opt. Since every node has a loop, and you could do this trick every time your walk hits a loop from the correct direction (with the same attributes as the stored value), you hit a loop the correct way usually in 1 move, you'd get this calculation in constant time. 

You store in each node the value of traveling the loop that node must possess forwards and the value for traveling that loop backwards. If you come in from an outward facing edge (rather than loop facing) it will know exactly the value you'd get for traversing the loop and where you'd end up (self). If you have non-sequential elements like you get a bonus only if you came from this one other rung, you could cache that path transversal data and do the trick only if it matches.

Though, you'd still have O(n) for the 2-opt reversal and O(n) for the walk of the path forward and O(n) for the walk of the path backwards. Even if you used basically a linked list to just link the edges rather store the path, you'd get 2-opt for constant time, but still need to calculate the value for the walk. The nodes within the reversed loop wouldn't need to be updated (they'd still know the value for hitting their loop forward and the value for hitting their loop backwards). But you'd have to update all the start and end facing nodes, as their values wouldn't correctly include the modified 2-opt modified values (as their loops will have changed).

So you'd have to update n/2 nodes each 2-opt performed. Even if you performed the 2-opt for free. And also performed the cached valuation for free, you'd still be stuck updating the values in O(n).

Which means  your best case is walking with no update needed. So you check n nodes if a 2-opt would help them and all of them wouldn't be helped and the checks would be in constant time (because we cached the values), O(n). In worst case you'd perform a 2-opt on every single node. So for the n nodes we'd need to update n values. That's O(n^2). So even with a bunch of tricks we still have O(n^2). However, in the subsequent passes, we'd still have the loop value information and wouldn't need to do anything between passes. So we could perform 2-opt at position 0 to position n then continue doing 2-opt at position 0 to position n, until things stop improving, but if we don't modify anything those checks are constant time. The nodes would already know the value of doing that loop forwards as well as doing it backwards, so you'd have the value for traversing the path done. You might only be able to cache things if they are non-cumulative though, but this would mean having the loop caches check for the incoming context of the path. But you could also just walk the entire path without that caching scheme, we're still talking O(n^2) for an entire pass.

Each 2-opt will take at most 3n operations (walk forward, walk backwards, perform the reversal). And a single pass will be n 2-opts. Admittedly you could do several passes until no operation can improve things.  Which means you'll be doing something like p * n 2-opts where p is number of passes. However I've done this algorithm before, it doesn't actually take much time, and you could likely get better results than you are currently getting in a single pass.

Even without tricks like caching the incoming looped path valuations, or using a data structure that allows 2-opt to be performed in constant time. You'd be able to get O(n^2) time in some pretty simple code. There's no higher exponent lurking in the shadows unless you somehow cannot give a value for particular path within n^2, which seems harsh since it seems like it'll be at worst linear time. There are some tricks that could improve the speed but they won't matter to the worst-case time complexity, only best and average.  It seems like it would be best to just implement the naive form without any tricks. It's not that many lines of code and pretty straight forward. And it's often no worse than doing a sort on the data.

I've done 2-opt on a hundred thousand stitches, where I also drew the entire pattern between each operation. It's perfectly reasonable. Also, your nodes aren't actually 1:1 with stitches. Most of those stitches are internal and have an order of 2. There's no 2-opt possible for that. You won't have an n of thousands, you will have an n of ~2*number_of_rungs.

**Comment by lexelby** [comment]
Hmm, alright then.  I'll give it a go sometime and we'll see what we get :)

**Comment by lexelby** [comment]
@carolundy Great news!  I was working on the auto-fill code for unrelated reasons, and the changes turned out to make underpathing easy to implement.  It's done!  You can find a test version here:

https://github.com/inkstitch/inkstitch/releases/tag/dev-build-lexelby-revamp-fill

We're discussing over in #409.  If you get a chance to test it, please let us know how it works for you, what you think of it, and if you find any problems.

**Comment by carolundy** [comment]
This is a great achievement. I did see the activity on #409 and was getting excited  that you were close to a solution. I am away from home teaching the Bernina software for the next few days, but will try to test when I get home. Thank you.
